src/python_testing/TC_CC_2_1.py

```diff
@@ -0,0 +1,400 @@
+
+        self.step(13)
+        await self._verify_attribute(self.attributes.EnhancedColorMode, ValueTypesEnum.ENUM, enum_range=range(0, 4))
+
+        self.step(14)
+        await self._verify_attribute(self.attributes.ColorLoopActive, ValueTypesEnum.UINT8)
+
+        self.step(15)
+        await self._verify_attribute(self.attributes.ColorLoopDirection, ValueTypesEnum.UINT8)
+
+        self.step(16)
+        await self._verify_attribute(self.attributes.ColorLoopTime, ValueTypesEnum.UINT16)
+
+        self.step(17)
+        await self._verify_attribute(self.attributes.ColorLoopStartEnhancedHue, ValueTypesEnum.UINT16)
+
+        self.step(18)
+        await self._verify_attribute(self.attributes.ColorLoopStoredEnhancedHue, ValueTypesEnum.UINT16)
+
+        self.step("18.a")
+        # read and save FeatureMap attribute
+        feature_map_value = await self.read_single_attribute_check_success(cluster=self.cluster, endpoint=self.endpoint, attribute=self.attributes.FeatureMap)
+
+        self.step(19)
+        color_capabilities_value = await self._verify_attribute(self.attributes.ColorCapabilities, ValueTypesEnum.UINT16, max_len=0x001f)
+        # verify the first 4 bits of colorcapabilities are the same on FeatureMap
+        self._verify_first_4bits(feature_map_value, color_capabilities_value)
+
+        self.step(20)
+        colortempphysicalminmireds_val = await self._verify_attribute(self.attributes.ColorTempPhysicalMinMireds, ValueTypesEnum.UINT16, min_len=1, max_len=0xfeff)
+
+        self.step(21)
+        colortempphysicalmaxmireds_val = await self._verify_attribute(self.attributes.ColorTempPhysicalMaxMireds, ValueTypesEnum.UINT16, min_len=1, max_len=0xfeff)
+
+        self.step(22)
+        await self._verify_attribute(self.attributes.CoupleColorTempToLevelMinMireds, ValueTypesEnum.UINT16, min_len=colortempphysicalminmireds_val, max_len=colortempphysicalmaxmireds_val)
+
+        self.step(23)
+        # Manual check
+        if await self.attribute_guard(endpoint=self.endpoint, attribute=self.attributes.StartUpColorTemperatureMireds):
+            sctmr_val = await self.read_single_attribute_check_success(cluster=self.cluster, endpoint=self.endpoint, attribute=self.attributes.StartUpColorTemperatureMireds)
+            asserts.assert_true(sctmr_val is None or (sctmr_val >= 1) and (sctmr_val <= 65279), "Value is out of range.")
+
+        # NumberOfPrimaries will be used to verify Primary<n>[X|Y|Intensity]
+        # Number of primaries cant be 0, it should be greater or equal than 1.
+        # Issue: #9103 to address remove 0 in NumberoOfPrimaries.
+        # After this is resolved, check not 0.
+
+        self.step(24)
+        # Read NumberOfPrimaries from the cluster.
+        numberofprimaries_value = await self._verify_attribute(self.attributes.NumberOfPrimaries, ValueTypesEnum.UINT8, min_len=0, max_len=6)
+        if numberofprimaries_value == 0:
+            logger.info("NumberOfPrimaries is 0 skipping from 25 to 43.")
```

Comment:
Do you want to skip 43? This `for step_num in range(25. 43)` will actually only skip 42 inclusive.

Reply:
Maybe you meant `Skipping steps 25 through 42`